### PR TITLE
feat: add support for localaddress

### PR DESCRIPTION
**Pull request by metcoder95**
## This relates to...

- #1578

## Changes

- feat: add support for localaddress

### Features

- Adds support for `localAddress` by forwarding the option to the `net.Socket#connect`  method.

## Status

KEY: S = Skipped, x = complete

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by metcoder95** [comment]
The PR is overall ready for review, I'm having a few struggles trying to test it as by default the `socket.localAddress` is `127.0.0.1`.
I tried with a DNS lookup to see if IPV6 is available and use it instead, but I'm facing errors on Undici when trying to do so.

Any good hint about how to test this properly? 🙂

**Comment by mcollina** [comment]
what error are you facing?
How is node.js testing for this feature?

**Comment by metcoder95** [comment]
<details>
<summary>`EINVAL ::1` for IPV6</summary>

 ✖ bind EINVAL ::1

  test: basic connect with localaddress
  at:
    line: 919
    column: 18
    file: node:net
    function: internalConnect

</details>


I can keep it with `127.0.0.1` but that wouldn't really be testing the implementation as this is the default address use when doing `socket.connect`.

>How is node.js testing for this feature?

Didn't check that yet, I'll do that tomorrow 👍

**Comment by mcollina** [comment]
> I can keep it with 127.0.0.1 but that wouldn't really be testing the implementation as this is the default address use when doing socket.connect.

Is it? `localhost` can be resolved to either `::1` or `127.0.0.1`.

**Comment by metcoder95** [comment]
> > I can keep it with 127.0.0.1 but that wouldn't really be testing the implementation as this is the default address use when doing socket.connect.
> 
> Is it? `localhost` can be resolved to either `::1` or `127.0.0.1`.

Yes, at least at my machine. The socket is automatically resolving to `127.0.0.1` and when trying to force the switch to `::1`, I automatically have the `EINVAL` error.

<details>
<summary>Sample</summary>
`Ignore the DNS, I was playing around to see if maybe the IPV6 was not being resolved locally`

```js
test('basic connect with localaddress', { only: true }, (t) => {
  t.plan(4)

  const server = http.createServer((c) => {
    t.fail()
  })
  server.on('connect', (req, socket, firstBodyChunk) => {
    socket.write('HTTP/1.1 200 Connection established\r\n\r\n')

    let data = firstBodyChunk.toString()
    socket.on('data', (buf) => {
      data += buf.toString()
    })

    socket.on('end', () => {
      socket.end(data)
    })
  })
  t.teardown(server.close.bind(server))

  dns.lookup('localhost', { family: 6 }, (err, address) => {
    if (err) t.fail(err)

    server.listen(0, async () => {
      const client = new Client(`http://localhost:${server.address().port}`, {
        // localAddress: '127.0.0.1'
      })
      t.teardown(client.close.bind(client))

      const signal = new EE()
      const promise = client.connect({
        signal,
        path: '/'
      })
      t.equal(signal.listenerCount('abort'), 1)
      const { socket } = await promise
      t.equal(signal.listenerCount('abort'), 0)
      t.equal(socket.localAddress, 'wrong-input')

      let recvData = ''
      socket.on('data', (d) => {
        recvData += d
      })

      socket.on('end', () => {
        t.equal(recvData.toString(), 'Body')
      })

      socket.write('Body')
      socket.end()
    })
  })
})
```


</details>

**Comment by mcollina** [comment]
Are you sure there is an IPv6 network on your machine?

**Comment by metcoder95** [comment]
It's enabled and set default to `automatic`. Not in use right now, but `::1` is properly linked to `localhost`.

So, not 100% how to test it

File: test/client-errors.js [hunk 1]
```diff
@@ -299,12 +299,22 @@ test('invalid options throws', (t) => {
 
   try {
     new Client(new URL('http://localhost:200'), { // eslint-disable-line
-      keepAliveTimeout: 'asd'
```
[the remark this answers is not in the document]
Comment (metcoder95) [reply test/client-errors.js]: Finger error, brought back in [82c2025](https://github.com/nodejs/undici/pull/1659/commits/82c20250664322f3c79d26edbf0abc6d94a39f16) 🙂

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**